File: scripts/smoke/ui.ts
/**
 * Browser smoke test.
 *
 * The existing smoke tests only exercise the APIs, so a purely client-side fault stayed
 * invisible: the industry picker treated TikTok's numeric `id` as a string, and every
 * visit to the ads page died with "a.id.slice is not a function" before rendering
 * anything. The APIs were all healthy throughout.
 *
 * This loads every page in a real browser, clicks the controls, and fails on any
 * uncaught exception or error-boundary screen.
 *
 *   npm run smoke:ui
 */
import { chromium, type Page } from 'playwright'

const BASE = process.env.BASE ?? 'http://localhost:3000'
const sleep = (ms: number) => new Promise((r) => setTimeout(r, ms))

let failures = 0
function check(label: string, ok: boolean, detail = '') {
  console.log(`${ok ? 'PASS' : 'FAIL'}  ${label}${detail ? ` — ${detail}` : ''}`)
  if (!ok) failures++
}

type Problem = { where: string; text: string }

/** Next.js renders this text when a client component throws during render. */
async function isErrorScreen(page: Page): Promise<boolean> {
  return page.evaluate(() =>
    /Application error|client-side exception|Unhandled Runtime Error/i.test(document.body.innerText),
  )
}

async function main() {
  const browser = await chromium.launch({ headless: true })
  const page = await browser.newPage({ viewport: { width: 1400, height: 950 } } as never)

  const problems: Problem[] = []
  let stage = 'boot'
  page.on('pageerror', (e) => problems.push({ where: stage, text: e.message }))
  page.on('console', (m) => {
    if (m.type() === 'error') problems.push({ where: stage, text: m.text() })
  })

  console.log('=== 1. Every page renders without a client exception ===')
  for (const [path, marker] of [
    ['/ads', 'Research quảng cáo'],
    ['/keywords', 'Research từ khoá'],
    ['/guide', 'Hướng dẫn sử dụng'],
  ] as const) {
    stage = `load ${path}`
    await page.goto(`${BASE}${path}`, { waitUntil: 'domcontentloaded', timeout: 60_000 })
    await sleep(3500)
    const broken = await isErrorScreen(page)
    const hasMarker = await page.evaluate((m) => document.body.innerText.includes(m), marker)
    check(`${path} renders`, !broken && hasMarker, broken ? 'error screen' : hasMarker ? '' : `missing "${marker}"`)
  }

  console.log('\n=== 2. Controls survive being clicked ===')
  stage = 'interact /ads'
  await page.goto(`${BASE}/ads`, { waitUntil: 'domcontentloaded', timeout: 60_000 })
  await sleep(3000)

  const chipCount = await page.locator('.chip').count()
  check('filter chips are present', chipCount > 0, `${chipCount} chips`)
  for (let i = 0; i < Math.min(chipCount, 16); i++) {
    await page.locator('.chip').nth(i).click().catch(() => {})
    await sleep(250)
    if (await isErrorScreen(page)) {
      const label = await page.locator('.chip').nth(i).innerText().catch(() => `#${i}`)
      check(`clicking chip "${label}" is safe`, false, 'error screen appeared')
      break
    }
  }
  check('no error screen after clicking chips', !(await isErrorScreen(page)))

  const boxes = page.locator('.check input')
  for (let i = 0; i < (await boxes.count()); i++) {
    await boxes.nth(i).click().catch(() => {})
    await sleep(250)
  }
  check('no error screen after toggling checkboxes', !(await isErrorScreen(page)))

  console.log('\n=== 3. TikTok industry picker ===')
  // It mounts only while TikTok is a selected source, and it is the component that broke.
  stage = 'industry picker'
  await page.goto(`${BASE}/ads`, { waitUntil: 'domcontentloaded', timeout: 60_000 })
  await sleep(2500)
  const picker = page.locator('select').filter({ hasText: 'ngành hàng' }).first()
  const pickerExists = (await page.locator('.field', { hasText: 'Ngành hàng' }).count()) > 0
  check('industry picker is rendered', pickerExists)

  if (pickerExists) {
    // Populating it needs a warmed TikTok browser session, so allow generous time.
    let options = 0
    for (let i = 0; i < 40; i++) {
      options = await page.locator('.field', { hasText: 'Ngành hàng' }).locator('option').count()
      if (options > 1) break
      await sleep(3000)
    }
    check('industry list populated', options > 1, `${options} options`)
    const groups = await page.locator('.field', { hasText: 'Ngành hàng' }).locator('optgroup').count()
    check('industries are grouped by parent category', groups > 1, `${groups} groups`)
    check('no error screen after picker loaded', !(await isErrorScreen(page)))
    void picker
  }

  console.log('\n=== 4. Uncaught errors observed ===')
  const unique = [...new Map(problems.map((p) => [p.text.slice(0, 140), p])).values()]
  for (const p of unique) console.log(`   [${p.where}] ${p.text.slice(0, 200)}`)
  check('no uncaught client errors', unique.length === 0, `${unique.length} distinct`)

  console.log(`\n${failures === 0 ? 'ALL CHECKS PASSED' : `${failures} CHECK(S) FAILED`}`)
  await browser.close()
  process.exit(failures === 0 ? 0 : 1)
}

main().catch((e) => {
  console.error('smoke-ui failed:', e)
  process.exit(1)
})

export {}

File: backend/scripts/smoke/ui.py
"""
Smoke test trình duyệt.

Các smoke test còn lại chỉ chạm tới API, nên một lỗi thuần phía client vẫn vô hình: ô chọn
ngành hàng từng coi `id` dạng số của TikTok là chuỗi, và mọi lần vào trang Quảng cáo đều
chết với "a.id.slice is not a function" trước khi vẽ được gì. Toàn bộ API lúc đó vẫn khoẻ.

Script này mở từng trang bằng trình duyệt thật, bấm các nút, và báo lỗi khi có exception
chưa bắt hoặc màn hình lỗi.

    python scripts/smoke/ui.py

BASE trỏ vào giao diện Next (cổng 3000), không phải backend.
"""

from __future__ import annotations

import asyncio
import os
import re
import sys

from playwright.async_api import Page, async_playwright

BASE = os.environ.get("BASE", "http://localhost:3000")

failures = 0


def check(label: str, ok: bool, detail: str = "") -> None:
    global failures
    print(f"{'PASS' if ok else 'FAIL'}  {label}{f' — {detail}' if detail else ''}")
    if not ok:
        failures += 1


ERROR_SCREEN = re.compile(
    r"Application error|client-side exception|Unhandled Runtime Error", re.IGNORECASE
)


async def is_error_screen(page: Page) -> bool:
    """Next.js in ra những câu này khi một client component ném lỗi lúc render."""
    text = await page.evaluate("() => document.body.innerText")
    return bool(ERROR_SCREEN.search(text or ""))


async def main() -> int:
    problems: list[tuple[str, str]] = []
    stage = "boot"

    async with async_playwright() as pw:
        browser = await pw.chromium.launch(headless=True)
        page = await browser.new_page(viewport={"width": 1400, "height": 950})

        page.on("pageerror", lambda e: problems.append((stage, str(e))))
        page.on(
            "console",
            lambda m: problems.append((stage, m.text)) if m.type == "error" else None,
        )

        print("=== 1. Mọi trang render được, không có exception phía client ===")
        for path, marker in [
            ("/ads", "Research quảng cáo"),
            ("/keywords", "Research từ khoá"),
            ("/guide", "Hướng dẫn sử dụng"),
        ]:
            stage = f"load {path}"
            await page.goto(f"{BASE}{path}", wait_until="domcontentloaded", timeout=60_000)
            await asyncio.sleep(3.5)
            broken = await is_error_screen(page)
            has_marker = await page.evaluate("(m) => document.body.innerText.includes(m)", marker)
            check(
                f"{path} render được",
                not broken and has_marker,
                "màn hình lỗi" if broken else ("" if has_marker else f'thiếu "{marker}"'),
            )

        print("\n=== 2. Các ô điều khiển chịu được việc bị bấm ===")
        stage = "interact /ads"
        await page.goto(f"{BASE}/ads", wait_until="domcontentloaded", timeout=60_000)
        await asyncio.sleep(3)

        chip_count = await page.locator(".chip").count()
        check("có chip bộ lọc", chip_count > 0, f"{chip_count} chip")
        for i in range(min(chip_count, 16)):
            try:
                await page.locator(".chip").nth(i).click()
            except Exception:
                pass
            await asyncio.sleep(0.25)
            if await is_error_screen(page):
                try:
                    label = await page.locator(".chip").nth(i).inner_text()
                except Exception:
                    label = f"#{i}"
                check(f'bấm chip "{label}" an toàn', False, "màn hình lỗi hiện ra")
                break
        check("không có màn hình lỗi sau khi bấm chip", not await is_error_screen(page))

        boxes = page.locator(".check input")
        for i in range(await boxes.count()):
            try:
                await boxes.nth(i).click()
            except Exception:
                pass
            await asyncio.sleep(0.25)
        check("không có màn hình lỗi sau khi tick checkbox", not await is_error_screen(page))

        print("\n=== 3. Ô chọn ngành hàng TikTok ===")
        # Nó chỉ mount khi TikTok đang được chọn làm nguồn, và chính nó là thứ từng hỏng.
        stage = "industry picker"
        await page.goto(f"{BASE}/ads", wait_until="domcontentloaded", timeout=60_000)
        await asyncio.sleep(2.5)
        field = page.locator(".field", has_text="Ngành hàng")
        picker_exists = await field.count() > 0
        check("ô chọn ngành hàng được render", picker_exists)

        if picker_exists:
            # Đổ dữ liệu vào nó cần một phiên trình duyệt TikTok đã làm nóng, nên chờ rộng tay.
            options = 0
            for _ in range(40):
                options = await field.locator("option").count()
                if options > 1:
                    break
                await asyncio.sleep(3)
            check("danh sách ngành hàng đã nạp", options > 1, f"{options} lựa chọn")
            groups = await field.locator("optgroup").count()
            check("ngành hàng được gom nhóm theo ngành cha", groups > 1, f"{groups} nhóm")
            check("không có màn hình lỗi sau khi ô chọn nạp xong", not await is_error_screen(page))

        print("\n=== 4. Lỗi chưa bắt quan sát được ===")
        unique = list({text[:140]: (where, text) for where, text in problems}.values())
        for where, text in unique:
            print(f"   [{where}] {text[:200]}")
        check("không có lỗi client chưa bắt", len(unique) == 0, f"{len(unique)} lỗi khác nhau")

        await browser.close()

    print(f"\n{'TẤT CẢ ĐỀU ĐẠT' if failures == 0 else f'{failures} MỤC KHÔNG ĐẠT'}")
    return 0 if failures == 0 else 1


if __name__ == "__main__":
    sys.stdout.reconfigure(encoding="utf-8")
    try:
        sys.exit(asyncio.run(main()))
    except Exception as e:  # noqa: BLE001
        print(f"smoke-ui chạy lỗi: {e}")
        sys.exit(1)
